orderer/common/cluster/comm.go

```diff
@@ -640,6 +684,70 @@ func (stream *Stream) operateWithTimeout(invoke StreamOperation, report func(err
 	}
 }
 
+func (stream *Stream) AuthenticateStream(version uint32, fromId uint64, toId uint64, signer identity.SignerSerializer) error {
+	if signer == nil {
+		return errors.New("signer is nil")
+	}
+
+	timestamp, err := ptypes.TimestampProto(time.Now().UTC())
+	if err != nil {
+		return errors.Wrap(err, "failed to read timestamp")
+	}
+
+	payload := &orderer.NodeAuthRequest{
+		Version:   version,
+		Timestamp: timestamp,
+		FromId:    fromId,
+		ToId:      toId,
+		Channel:   stream.Channel,
+	}
+
+	bindingFieldsHash := GetSessionBindingHash(payload)
+
+	tlsBinding, err := GetTLSSessionBinding(stream.Cluster_StepClient, bindingFieldsHash)
+	if err != nil {
+		return errors.Wrap(err, "TLSBinding failed")
+	}
+	payload.SessionBinding = tlsBinding
+
+	asnSignFields, _ := asn1.Marshal(AuthRequestSignature{
+		Version:   int64(payload.Version),
+		Timestamp: payload.Timestamp.String(),
+		FromId:    strconv.FormatUint(payload.FromId, 10),
+		ToId:      strconv.FormatUint(payload.ToId, 10),
+		Channel:   payload.Channel,
+	})
+	sig, err := signer.Sign(asnSignFields)
+	if err != nil {
+		return errors.Wrap(err, "signing failed")
+	}
+
+	payload.Signature = sig
+	stepRequest := &orderer.ClusterNodeServiceStepRequest{
+		Payload: &orderer.ClusterNodeServiceStepRequest_NodeAuthrequest{
+			NodeAuthrequest: payload,
+		},
+	}
+
+	f := func() (*orderer.StepResponse, error) {
+		cli := stream.Cluster_StepClient.(orderer.ClusterNodeService_StepClient)
+		err := cli.Send(stepRequest)
+		return nil, err
+	}
+
+	_, err = stream.operateWithTimeout(f, func(_ error) {})
+	if err != nil {
+		return errors.Wrapf(err, "stream authentication request failed")
+	} else {
+		_, err = stream.Recv()
```

Comment:
what do you expect to receive back?

Reply:
while authenticating to server, it expects a response to check whether authentication is successful or not

Reply:
But that's not the server side code is doing though.... where do you do that in the server? 

And why send a request back? If you're not authorized, the server can just close the stream with an error :-)